ext/opentelemetry-ext-dbapi/src/opentelemetry/ext/dbapi/__init__.py

```diff
@@ -0,0 +1,140 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+The opentelemetry-ext-dbapi package allows tracing queries made by the
+ibraries following Ptyhon Database API specification:
+https://www.python.org/dev/peps/pep-0249/
+"""
+
+import typing
+
+import wrapt
+
+from opentelemetry.trace import SpanKind, Tracer
+from opentelemetry.trace.status import Status, StatusCanonicalCode
+
+QUERY_WRAP_METHODS = ["execute", "executemany", "callproc"]
+
+
+def trace_integration(
+    tracer: Tracer,
+    connect_module: typing.Callable[..., any],
+    connect_method_name: str,
+    database_component: str,
+):
+    """Integrate with DB API library.
+       https://www.python.org/dev/peps/pep-0249/
+    """
+    # pylint: disable=unused-argument
+    def wrap_connect(
+        wrapped: typing.Callable[..., any],
+        instance: typing.Any,
+        args: typing.Tuple[any, any],
+        kwargs: typing.Dict[any, any],
+    ):
+        db_integration = DatabaseApiIntegration(tracer, database_component)
+        return db_integration.wrapped_connection(wrapped, args, kwargs)
+
+    wrapt.wrap_function_wrapper(
+        connect_module, connect_method_name, wrap_connect
+    )
+
+
+class DatabaseApiIntegration:
+    def __init__(
+        self, tracer: Tracer, database_component: str, database_type="sql"
+    ):
+        if tracer is None:
+            raise ValueError("The tracer is not provided.")
+        self._tracer = tracer
+        self._database_component = database_component
+        self._database_type = database_type
+        self._connection_props = {}
+
+    def wrapped_connection(
+        self,
+        wrapped: typing.Callable[..., any],
+        args: typing.Tuple[any, any],
+        kwargs: typing.Dict[any, any],
+    ):
+        """Patch connection cursor to add tracing.
+        """
+        connection = wrapped(*args, **kwargs)
+        self._connection_props = {
+            "database": connection.database,
+            "port": connection.server_port,
+            "host": connection.server_host,
+            "user": connection.user,
+        }
+
+        wrapt.wrap_function_wrapper(connection, "cursor", self.wrapped_cursor)
+        return connection
+
+    # pylint: disable=unused-argument
+    def wrapped_cursor(
+        self,
+        wrapped: typing.Callable[..., any],
+        instance: typing.Any,
+        args: typing.Tuple[any, any],
+        kwargs: typing.Dict[any, any],
+    ):
+        """Patch cursor execute, executemany and callproc methods.
+        """
+        cursor = wrapped(*args, **kwargs)
+        for func in QUERY_WRAP_METHODS:
+            wrapt.wrap_function_wrapper(cursor, func, self.add_span)
```

Comment:
Is it possible for any of these to throw an exception while attempting to wrap? If so, should we abort the wrap for all of them?

Reply:
Do we want to integrate with some methods but not all of them?, sound better to just leave the exception to raise instead of try/except here